### PR TITLE
Schedule and Reminder notification

### PR DESCRIPTION
**Pull request by bigvaimarcella**
# Description
**Reminder Feature:**  [Wiki](https://bcdevex.atlassian.net/wiki/spaces/CCP/pages/1133838349/Reminder+feature)

 

This feature allows a form designer to be able to enable automated submission reminders at the form level for a given submission period. Once the submission period is set, the option to enable automated notifications will be enabled for users to choose to use it or not.

 

If checked on, the system will send 3 notifications:

 

1 at the start of the submission period
1 halfway through the submission period
1 a day prior to the submission period

 

Reminders would only be sent to the identified “Submitters“ who haven’t submitted the data for the given period.

 

A generic template will be used for the notifications. The content should always include the link to the form. Future possible enhancements (post-MVP):

 

Have a dedicated notification section to allow for more fine-tuning of the functionality

 

Priority: High (not critical but easy to add to the scope of the MVP)

 

Health Authorities have to notify operators manually every submission period to let them know they must submit their data on time.

 

Each submission period is 45 days long. They send a notification at the start, then send reminders to the people who haven’t submitted until the end of the period.

 

This is very time-consuming as they do not have an easy way to track submissions. They have to prepare the email with the correct recipients making sure they’re not sending it to someone who’s already submitted it.

 

How to use it

 

This feature’s development has been made in three parts, first the frontend part, next the back-end, and the last one the cronjob.

 

In the front end, after creating a form save it and go to this form management, after publishing a version of it, the Schedule setting checkbox is enabled, now you can check it, after doing that the section FORM SCHEDULE SETTINGS appear. it will contain a list of inputs to configure the period.

 

First, we have the open submission date input as the starting date of the period.

 

We 3 types of periods represented by 3 radios button

 

One with no end date so we can close at any time

 

One with a specific date to end

 

The last type we can set up a submission period
You have to choose one to continue.

 

Accord the type you choose, you can see a checkbox to active late submission when we check it we have 2 new inputs one for the term and the other for the interval.

 

NB: we only see this checkbox for late submission when we choose the second and the third type.

 

If you check the last type of period 2 new inputs appear and a checkbox for the repeat period.

 

With the 2 inputs, we can config the during of the period one input for the term and the other for the interval.

 

And the checkbox is to active repeat period, check it, 3 new fields appear to config the repeat period according to the frequency that we want and the last input is for the end date.

 

We have a checkbox for active customs closing messages when we check it we have a new text area for that.

 

The last checkbox allows enabling reminder submission, when you check it you are active now the mail reminder all submitters of this form will receive emails from the reminder on the opening day of the form, in the middle if they don’t yet, and a last one the day before the form close. NB those emails depend also on the type period by example for the first type that one without closed we send only one mail the first one.

 

Reminder notification is a service of chefs that allows the system to send reminder emails to the submitter of a specific form during the open submission period. Generally, it sends three emails, one on the first day of opening, the second in the middle, and the last one day before the end of the period.
We have three types of form schedules that trigger reminder emails:

 

form without an End Date

 

form with a Start and End Date

 

form with a Repeat Section
When the form has no end date, we send only one email the first, but if the date is after the current date.

 

When the form has a start and end date or a repeat period, we send three emails, but according to the length of the period. For example, if the date is less than three days, we send only one email the first. If it is between 4-6 days, we send two emails, the first and the last. If the period is more than six days, we send three emails, one at the start, one in the middle and the day before the end of the period.

 

What is a period?It is a moment when a form is open for submission, containing a start date and an end date. There are two types of periods. A simple one includes only a start and end date. The second one also has a start and end date but repeats according to an interval.

 

Technical Details:
Open Shift can be responsible to handle Schedule/Cron Job (In order to send Bulk Emails). The cronjob triggers every day at 8 AM O’clock

 

Common Components - Digital Government - Province of British Columbia (Common Hosted Email Service) can be used to send bulk emails.

 

CHES Service file : app\src\components\chesService.js

 

Use of chesService : app\src\forms\email\emailService.js

 

API Documentation Common Hosted Email Service API - Documentation v1 (http://gov.bc.ca/ )

 

Need multiple Login accounts for testing

 

In order to create multiple accounts, We need to enable login by BCeID (Identity provider) module. Here is rocket chat help thread BCDevXHQ for this.

 

In order to create (Bypass Identity providers) local users need some changes into KeyCloak admin console. (More info in the document attached)

 

**Schedule Feature:** [Wiki](https://bcdevex.atlassian.net/wiki/spaces/CCP/pages/1143242777)
This feature allows the form designer to schedule form submissions. By enabling this feature form designer can Limit the submissions to a form by selecting Open and Close submission dates or a specific period. The schedule feature can be enabled by clicking on the "Form Submission Schedule" checkbox under the Form schedule settings on form settings. One thing to remember is that this feature can only be available to the forms that have at least one version published. Three scenarios can schedule a form submission.

 

By enabling this radio box "Keep open until manually unpublished" the form submission will work normally, the only limit a form designer can use is to select an "Open submissions date" so basically form will be open on a specific date and will be available until the form get unpublished manually.

 

"Schedule a closing date" is another option where the form designer can select a specific closing date for a form submission. Selecting the Open and Close date form will open submissions for a specific selected period only. Plus, the form designer can allow end users to submit the form with a Grace period. To do so, the Form designer can use the "Allow late submissions" option and choose a specific interval.
Upon closing a form, the Form designer can set a custom message for the end user as well. That can be seen on the front end once the form gets closed for a specific period. This feature can be used by clicking on the "Set custom closing message" checkbox.

 

Last option "Set up submission period" will act similarly to the previous one but setting up this option is a bit different. Here Form designer has to choose a period and interval to close the form submission. The first input accepts a Positive integer/number and the next one accepts a period that could be Days, Weeks, Months, etc. So form will be open from the Open submission date to the Period selected by the form designer. Again form designers can allow end users to give a grace period for submission by enabling allow late submission option.
One advanced feature here is to set up a Repeat period, Where a Form Designer can set up a form availability repeatedly. By means, the form designer can select a particular repetition to occur for the current form by inputting a positive number in the very first input box that says "Every" as a placeholder, then using the second drop box to select a period, and the last thing using an End date when the repetition should end.
Upon closing a form, the Form designer can set a custom message for the end user as well. That can be seen on the frontend once the form gets close for a specific period. This feature can be used by clicking on the "Set custom closing message" checkbox.

 

**Copy from existing submission**: [Wiki](https://bcdevex.atlassian.net/wiki/spaces/CCP/pages/1142849541/Copy+from+existing+submission)
This feature allows submitters to create new submissions from previous ones. By creating new submissions from existing ones, the user can copy all the previous values for the fields enabled with “Allow value propagation”. All the form fields enabled with this option can be copied over to new submission.

 

This is how a Form designer can Enable this feature from Form Functionality.

 

ThisIsHow

 

Next, Enable the “Allow value propagation” checkbox for each field you want to copy while the submitter creates new submission from the old one. Designer can enable this option for fields under the validation tab of Field management while Designing a form. Remember that This option can only be available for Advanced components only. i.e.

 

Example

 

Now Submitters can see a new action button under My Submissions table called “Copy this submission”. Like this:

 

Previous

 

By clicking on that Icon user will be redirected to the New page where all the existing values from the older submission are auto-populated to the form.

 

    We replaced reminder by reminder_enabled now its a boolean and adding documentation in API specification for reminder_enabled and schedule fields.

 

## Types of changes

 

 

New feature (non-breaking change which adds functionality)

 

## Checklist

 

 

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have run the npm script lint on the frontend and backend
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have approval from the product owner for the contribution in this pull request

### CODE REVIEW

File: app/src/forms/email/emailService.js [hunk 1]
```diff
@@ -112,6 +113,67 @@ const buildEmailTemplate = async (formId, formSubmissionId, emailType, referer,
   };
 };
 
+/** Helper function used to build the email template based on email type and contents for reminder */
+const buildEmailTemplateFormForReminder = async (form, emailType, users, report, referer) => {
+  let configData = {};
+  const closeDate = moment(report.dates.closeDate).format('MMM. D, YYYY');
+  const subject = 'CHEFS Submission Reminder';
+  const message = (report.dates.closeDate) ? `This email is to inform you that the ${form.name} form is now open for submissions and will stay open until ${closeDate}. Please ensure to complete your submission before the period is closed.
+  Thank you.` : `This email is to inform you that the ${form.name} form is now open for submissions.
```
Comment (jasonchung1871) [inline app/src/forms/email/emailService.js:122]: This should say `This email is to inform you that the ${form.name} form is now open for submission.`

File: app/src/forms/email/emailService.js [hunk 2]
```diff
@@ -112,6 +113,67 @@ const buildEmailTemplate = async (formId, formSubmissionId, emailType, referer,
...
+      priority: 'normal',
+      form,
+    };
+  } else if (emailType === EmailTypes.REMINDER_FORM_NOT_FILL) {
+    configData = {
+      bodyTemplate: 'reminder-form-not-fill.html',
+      title: `Submission Reminder for ${form.name}`,
+      subject: subject,
+      messageLinkText: `Hi,\n
+      This email is to remind you that the ${form.name} form is open for submissions until ${ closeDate }. Please ensure to complete your submission before the period is closed.
```
Comment (jasonchung1871) [inline app/src/forms/email/emailService.js:141]: The message should say `This email is to inform you that the ${form.name} form is now open for submission and will stay open until ${closeDate}. Please complete your submission before the submission period is closed.`

File: app/src/forms/email/reminderService.js [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+            chesResponses
+          });
+          // eslint-disable-next-line no-unused-vars
+          mail+= result.submitters.length;
+        } else{
+          errors.push(q[i].message);
+        }
+      }
+      return {
+        messsage: `${q.length} Forms found , ${mail} mails sent.`,
```
Comment (jasonchung1871) [inline app/src/forms/email/reminderService.js:43]: This should remain consistent `${q.length} forms found, ${mail} emails sent.`

Also, message is spelled incorrectly, it should be `message` and not `messsage`

File: app/src/forms/email/reminderService.js [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    const resolve = [];
+    const errors = [];
+    let mail = 0;
+    
+    if(q.length!==0) {
+      for(let i = 0 ; i < q.length ; i++) {
+        if(!q[i].error){
+          const obj = await  service._runQueries(q[i].statement);
+          let result = await service._initStatement(obj);
+          let chesResponses = service._initMaillSender(result, referer);
```
Comment (jasonchung1871) [inline app/src/forms/email/reminderService.js:27]: This should be `service._initMailSender(result, referer);`

File: app/src/forms/email/reminderService.js [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+      }
+      return {
+        messsage: `${q.length} Forms found , ${mail} mails sent.`,
+        errors,
+        resolve
+      };
+    }
+    else {
+      return {
+        messsage: '0 email sent',
```
Comment (jasonchung1871) [inline app/src/forms/email/reminderService.js:50]: Message is spelled incorrectly, it should be `message` and not `messsage`.

The message itself should also say `0 emails sent.`

File: app/src/forms/email/reminderService.js [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+      });
+    return fs;
+  },
+  _getReminders : async (forms)=>{
+    let reminder = [];
+    let toDay = moment();
+    for (let i = 0; i< forms.length; i++) {
+      let obj = {};
+
+      obj.avalaibleDate =  service._listDates(forms[i].schedule) ;
```
Comment (jasonchung1871) [inline app/src/forms/email/reminderService.js:154]: availableDate is spelled incorrectly, it should be `obj.availableDate`

File: app/src/forms/email/reminderService.js [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+      if (obj.avalaibleDate.length == 0) {
+        reminder.push({ error:true, message : `Form ${forms[i].name } has no avalaible date.` });
```
Comment (jasonchung1871) [inline app/src/forms/email/reminderService.js:157]: available is spelled incorrectly, it should be `available`

File: app/src/forms/email/reminderService.js [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+        continue;
+      }
+
+      obj.report = service.getCurrentPeriod(obj.avalaibleDate, toDay, forms[i].schedule.allowLateSubmissions.enabled);
```
Comment (jasonchung1871) [inline app/src/forms/email/reminderService.js:161]: availableDate is spelled incorrectly, it should be `obj.availableDate`

File: app/src/forms/email/reminderService.js [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+
+      obj.form   = forms[i];
+
+
+      obj.state = service._getMailType(obj.report, forms[i].schedule.allowLateSubmissions.enabled);
+
+      if(obj.state == undefined) {
+        reminder.push({ error:true, message : ` Form ${forms[i].name } has no valid date ` });
```
Comment (jasonchung1871) [inline app/src/forms/email/reminderService.js:169]: There shouldn't be a space at the beginning and end of the sentence.

File: app/src/forms/email/reminderService.js [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+      const host = process.env.SERVER_HOST;
+      return`${protocol}${host}${basePath}`;
+    } catch (error) {
+      log.error(error.message, {
+        function: '_getReferer'
+      });
+      throw error;
+    }
+  },
+  _initMaillSender: async (statement, referer) => {
```
Comment (jasonchung1871) [inline app/src/forms/email/reminderService.js:287]: This should be `_initMailSender: async(statement, referer) => {`

File: app/src/forms/form/service.js [hunk 11]
```diff
@@ -61,6 +67,9 @@ const service = {
       obj.enableStatusUpdates = data.enableStatusUpdates;
       obj.enableSubmitterDraft = data.enableSubmitterDraft;
       obj.createdBy = currentUser.usernameIdp;
+      obj.schedule = data.schedule;
+      obj.reminder_enabled = data.reminder_enabled ;
```
Comment (jasonchung1871) [inline app/src/forms/form/service.js:71]: There's a random space at the end before the semi colon.

File: openshift/app.cm.yaml
Comment (jasonchung1871) [inline openshift/app.cm.yaml:1]: What's the purpose of this template?

**Review by jasonchung1871 — CHANGES_REQUESTED** [review]

File: app/frontend/src/store/modules/form.js [hunk 12]
```diff
@@ -320,7 +357,10 @@ export default {
             userType: state.form.userType,
           }),
           showSubmissionConfirmation: state.form.showSubmissionConfirmation,
-          submissionReceivedEmails: emailList
+          submissionReceivedEmails: emailList,
+          schedule: schedule,
+          reminder_enabled : state.form.reminder_enabled,
```
[the remark this answers is not in the document]
Comment (bigvaimarcella) [reply app/frontend/src/store/modules/form.js:362]: I remove it

File: app/src/forms/email/emailService.js [hunk 13]
```diff
@@ -112,6 +113,67 @@ const buildEmailTemplate = async (formId, formSubmissionId, emailType, referer,
   };
 };
 
+/** Helper function used to build the email template based on email type and contents for reminder */
+const buildEmailTemplateFormForReminder = async (form, emailType, users, report, referer) => {
+  let configData = {};
+  const closeDate = moment(report.dates.closeDate).format('MMM. D, YYYY');
+  const subject = 'CHEFS Submission Reminder';
+  const message = (report.dates.closeDate) ? `This email is to inform you that the ${form.name} form is now open for submissions and will stay open until ${closeDate}. Please ensure to complete your submission before the period is closed.
```
[the remark this answers is not in the document]
Comment (bigvaimarcella) [reply app/src/forms/email/emailService.js:121]: I replace the message by the message you provide.

File: app/src/forms/email/emailService.js (hunk 2 above)
Comment (bigvaimarcella) [reply app/src/forms/email/emailService.js:141]: Text replace

File: app/frontend/src/components/designer/FormViewer.vue [hunk 14]
```diff
@@ -177,14 +209,35 @@ export default {
       return `Bearer ${this.token}`;
     },
     async getFormData() {
+      function deleteFieldData(fieldcomponent,submission) {
+        if(Object.prototype.hasOwnProperty.call(fieldcomponent,'components')){
+          fieldcomponent.components.map((subComponent) => { // Check if it's a Nested component
```
[the remark this answers is not in the document]
Comment (ptugger) [reply app/frontend/src/components/designer/FormViewer.vue:214]: I have already changed that and tested it with nested components. See this function deleteFieldData(subComponent,submission) that make it happen

File: app/frontend/src/components/designer/FormViewer.vue [hunk 15]
```diff
@@ -177,14 +209,35 @@ export default {
...
-        this.version = response.data.version.version;
+        if(!this.isDuplicate){
+          //As we know this is a Submission from existing one so we will wait for the latest version to be set on the getFormSchema
+          this.formSchema = response.data.version.schema;
+          this.version = response.data.version.version;
+        }else{
+          /** Let's remove all the values of such components that are not enabled for Copy existing submission feature */
+          if(response.data?.version?.schema?.components && response.data?.version?.schema?.components.length){
+            // console.log('response.data.version.schema.components--',response.data.version.schema.components);
+            response.data.version.schema.components.map((component) => {
```
[the remark this answers is not in the document]
Comment (ptugger) [reply app/frontend/src/components/designer/FormViewer.vue:236]: It should find it, as the same function gets looped to do it. But still let me test it once again

File: app/src/forms/email/reminderService.js (hunk 8 above)
Comment (ptugger) [reply app/src/forms/email/reminderService.js:161]: It was done last week

File: app/src/forms/email/emailService.js [hunk 16]
```diff
@@ -112,6 +113,67 @@ const buildEmailTemplate = async (formId, formSubmissionId, emailType, referer,
...
+      priority: 'normal',
+      form,
+    };
+  } else if (emailType === EmailTypes.REMINDER_FORM_WILL_CLOSE) {
+    configData = {
+      bodyTemplate: 'reminder-form-will-close.html',
+      title: `Submission Closing for ${form.name}`,
+      subject: subject,
+      messageLinkText: `Hi,\n
+      This email is to remind you that the ${form.name} form is open for submissions until ${ closeDate }. Please ensure to complete your submission before the period is closed.
```
[the remark this answers is not in the document]
Comment (bigvaimarcella) [reply app/src/forms/email/emailService.js:153]: I did not create a const for that cause message was not the same in the begining, now I consider your advice :
  ```
const formatEmailTextMessage = (name, closeDate )=> {
  const messageValue =  'Hi,\n'+(closeDate) ? `This email is to inform you that the ${name} form is now open for submission and will stay open until ${closeDate}. Please complete your submission before the submission period is closed.` : `This email is to inform you that the ${form.name} form is now open for submission.`;
    return messageValue+'\n Thank you';
  };
```
 const message = formatEmailTextMessage(form.name, closeDate);
I create one function for all mail to generate the message

File: app/src/forms/email/reminderService.js (hunk 6 above)
Comment (bigvaimarcella) [reply app/src/forms/email/reminderService.js:154]: Correct now its availableDate

File: app/frontend/src/components/designer/FormViewer.vue (hunk 14 above)
[the remark this answers is not in the document]
Comment (jasonchung1871) [reply app/frontend/src/components/designer/FormViewer.vue:214]: Sorry, I'm not entirely sure exactly where all nested components go but they don't all go in a variable called components. I believe some of the advanced layouts store them in a different manner.

File: app/src/forms/email/reminderService.js (hunk 8 above)
Comment (bigvaimarcella) [reply app/src/forms/email/reminderService.js:161]: available is spelled incorrectly, it should be available

File: app/src/forms/email/reminderService.js (hunk 7 above)
Comment (bigvaimarcella) [reply app/src/forms/email/reminderService.js:157]: Correct

File: app/src/forms/email/reminderService.js (hunk 9 above)
Comment (bigvaimarcella) [reply app/src/forms/email/reminderService.js:169]: Space remove

File: app/src/forms/email/reminderService.js (hunk 5 above)
Comment (jasonchung1871) [reply app/src/forms/email/reminderService.js:50]: This is an English grammar reason, see https://www.bbc.co.uk/worldservice/learningenglish/grammar/learnit/learnitv354.shtml

File: app/src/forms/email/reminderService.js (hunk 10 above)
Comment (bigvaimarcella) [reply app/src/forms/email/reminderService.js:287]: correct now

File: app/src/forms/email/reminderService.js (hunk 8 above)
Comment (jasonchung1871) [reply app/src/forms/email/reminderService.js:161]: It seems to have been added back in then

File: app/src/forms/form/service.js (hunk 11 above)
Comment (bigvaimarcella) [reply app/src/forms/form/service.js:71]: Space remove

File: app/src/forms/email/reminderService.js (hunk 3 above)
Comment (bigvaimarcella) [reply app/src/forms/email/reminderService.js:43]: Solve

File: app/frontend/src/components/designer/FormViewer.vue (hunk 14 above)
[the remark this answers is not in the document]
Comment (jasonchung1871) [reply app/frontend/src/components/designer/FormViewer.vue:214]: I would write a unit test that checks against different kinds of nested components that exist in CHEFS. I'm not entirely sure what the purpose of this function is, but would it break or would anything go wrong if it didn't delete some field components?

File: app/src/forms/email/reminderService.js (hunk 4 above)
Comment (bigvaimarcella) [reply app/src/forms/email/reminderService.js:27]: Solve

File: openshift/app.cm.yaml
Comment (bigvaimarcella) [reply openshift/app.cm.yaml:1]: Only @evgeny-pluzhnikov  can answer you for now

File: app/src/forms/email/reminderService.js (hunk 9 above)
Comment (bigvaimarcella) [reply app/src/forms/email/reminderService.js]: Solve

File: app/src/forms/email/reminderService.js (hunk 10 above)
Comment (bigvaimarcella) [reply app/src/forms/email/reminderService.js]: I solved it

File: openshift/app.cm.yaml
Comment (evgeny-pluzhnikov) [reply openshift/app.cm.yaml:1]: The purpose of this template is to accumulate all ConfigMap objects used by CHEFS so all of them can be provisioned during pipeline execution. The immediate need was to create ConfigMap to hold SERVER_HOST variable. It is used by mail notification service when compiling a link for some form submission.

File: app/frontend/src/components/designer/FormViewer.vue (hunk 14 above)
[the remark this answers is not in the document]
Comment (ptugger) [reply app/frontend/src/components/designer/FormViewer.vue:214]: Done and pushed

File: app/frontend/src/components/designer/FormViewer.vue (hunk 15 above)
[the remark this answers is not in the document]
Comment (ptugger) [reply app/frontend/src/components/designer/FormViewer.vue:236]: Solution pushed to Branch, Asgard will merge this asap

**Pull request closed without merging**